### ISSUE TITLE
Enable discovery of locally installed Typescript Language Service plugins

### ISSUE THREAD

**Issue by killerDJO** [Microsoft/vscode#45856]
Issue Type: <b>Feature Request</b>

Currently, the typescript language service (tsserver) used by VS Code doesn't discover language service plugins installed locally. Instead, it tries to search them relative to a typescript sdk location.
 
Steps to reproduce:
1. Specify a globally installed typescript version to use for a workspace (in general, typescript must be installed outside of the local project).
2. Install a typescript language service plugin to local node modules (e. g.  tslint-language-service).
3. Add language service plugin in tsconfig, e. g. 
"plugins": [ { "name": "tslint-language-service" } ]

Expected: Plugin is discovered and loaded.
Actual: Plugin is not discovered. TsServer tries to load it only relative to the typescript installation path.

Proposed change:
There is an option available in tsserver (--allowLocalPluginLoads) which enables discovery of locally installed plugins (available starting from TypeScript 2.3.3). Visual Studio already uses this option by default for TypeScript editing. The ideal solution would be to provide an option in VS Code to enable such behaviour (e. g. "typescript.tsserver.allowLocalPluginLoads": true) or enable it by default.

VS Code version: Code 1.21.0 (9a199d77c82fcb82f39c68bb33c614af01c111ba, 2018-03-07T11:04:09.886Z)
OS version: Windows_NT ia32 10.0.16299

**Comment by mjbvz** [comment]
This is by design for security reasons. I'll sync with the TS team to see what their thoughts are. At the very least, the option to load local plugins must be something that only a user can configure

**Comment by mjbvz** [comment]
Ok, this flag is a workaround and we shouldn't enable in vscode (even with a setting). Two options:

* Do nothing. We already load local plugins if you use the workspace version of TS. You can also always use plugins that are contributed by vs code extensions
* Add a `typescript.server.pluginPaths` user setting that lets people explicitly opt in a path

I'm in favor of option one unless there is a very good reason why it would not work

**Comment by killerDJO** [comment]
There are a couple of issues with usage of the workspace version of TS in case we have multiple TS projects in a single workspace.
With multiple projects, a good approach would be to install typescript in a common/main project and use it from there. However, this would force us to install all plugins in this project as well, even when only a single project requires this plugin.
Another annoyance it brings is that the same project can not be used homogeneously in VS Code and Visual Studio, since Visual Studio will try to use it's own typescript installation instead of the workspace one (it will also try to load project-local plugins since allowLocalPluginLoads flag is used). I understand this is a pretty rare case when we want to use the same project in both IDEs, but it's useful in large distributed teams when choice of IDE is not enforced.
Plugins from VS Code extensions, can be used to mitigate this to some extent, but this approach limits us to available extensions and forces us to use extensions for functionality that is provided by TS out of box.
I think for such cases (multiple TS projects per workspace) the ideal solution would be to provide an option to explicitly specify plugins search paths (`typescript.tsserver.pluginPaths`). In this case a user will not be enforced to use workspace version of TS (or extensions) and will be able to install plugins in any location he wants (whether globally for all projects or locally for single project). Since this option is opt in, it shouldn't bring any security issues and should be really helpful in advanced configuration scenarios because user will have full control over plugins installation location.

**Comment by killerDJO** [comment]
I've updated the `PR`. The only tricky part is about how relative paths are resolved. I've used the same approach as for the `typescript.tsdk`:
- if path starts with any workspace folder name, it's treated as relative to this folder;
- otherwise, path is resolved against each folder in a workspace.
This is consistent with the existing approach and supports pretty sophisticated configurations if required, e.g. project-local and workspace-global plugin paths can be specified in a single config.
